src/core/ext/transport/chaotic_good/server_transport.cc

```diff
@@ -328,27 +338,30 @@ auto ChaoticGoodServerTransport::OnTransportActivityDone(
 }
 
 ChaoticGoodServerTransport::ChaoticGoodServerTransport(
-    const ChannelArgs& args, std::unique_ptr<PromiseEndpoint> control_endpoint,
-    std::unique_ptr<PromiseEndpoint> data_endpoint,
-    std::shared_ptr<grpc_event_engine::experimental::EventEngine> event_engine)
+    const ChannelArgs& args, PromiseEndpoint control_endpoint,
+    PromiseEndpoint data_endpoint,
+    std::shared_ptr<grpc_event_engine::experimental::EventEngine> event_engine,
+    HPackParser hpack_parser, HPackCompressor hpack_encoder)
     : outgoing_frames_(4),
-      transport_(std::move(control_endpoint), std::move(data_endpoint)),
       allocator_(args.GetObject<ResourceQuota>()
                      ->memory_quota()
-                     ->CreateMemoryAllocator("chaotic-good")),
-      event_engine_(event_engine),
-      writer_{MakeActivity(TransportWriteLoop(),
-                           EventEngineWakeupScheduler(event_engine),
-                           OnTransportActivityDone("writer"))},
-      reader_{nullptr} {}
+                     ->CreateMemoryAllocator("chaotic-good")) {
+  auto transport = MakeRefCounted<ChaoticGoodTransport>(
+      std::move(control_endpoint), std::move(data_endpoint),
+      std::move(hpack_parser), std::move(hpack_encoder));
+  writer_ = MakeActivity(TransportWriteLoop(transport),
+                         EventEngineWakeupScheduler(event_engine),
+                         OnTransportActivityDone("writer"));
+  reader_ = MakeActivity(TransportReadLoop(std::move(transport)),
+                         EventEngineWakeupScheduler(event_engine),
+                         OnTransportActivityDone("reader"));
```

Comment:
When TransportWriteLoop is called, the refcount of transport will be incremented once (since transport is passed to it by value). Inside TransportWriteLoop, another std::move(transport) is made. - this shouldn't change the ref-count. In line 355, since `transport` is a local variable to this function and it hasn't been moved before, (it was only passed by value to TransportWriteLoop), it will be ok to call std::move(.transport)